File: 865/Solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* subtreeWithAllDeepest(TreeNode* root) {
        return helper(root).node;
    }

private:
    struct Result {
        TreeNode* node;
        int depth;

        Result(TreeNode* node, int depth) : node(node), depth(depth) {}
    };

    Result helper(TreeNode* node) {
        if (node == nullptr) {
            return Result(nullptr, 0);
        }

        Result left = helper(node->left);
        Result right = helper(node->right);

        if (left.depth > right.depth) {
            return Result(left.node, left.depth + 1);
        } else if (right.depth > left.depth) {
            return Result(right.node, right.depth + 1);
        } else {
            return Result(node, left.depth + 1);
        }
    }
};

File: 865/Solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def subtreeWithAllDeepest(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        return self.dfs(root, 0)[1]

    def dfs(self, node, depth):
        if not node:
            return depth, None

        left_depth, left_node = self.dfs(node.left, depth + 1)
        right_depth, right_node = self.dfs(node.right, depth + 1)

        if left_depth > right_depth:
            return left_depth, left_node
        elif right_depth > left_depth:
            return right_depth, right_node
        else:
            return left_depth, node